cpp/src/parquet/encoding-test.cc

```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  ::arrow::internal::ChunkedBinaryBuilder builder(static_cast<int>(dict_buffer_->size()),
+                                                  default_memory_pool());
+  auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder_.get());
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values =
+      byte_array_decoder->DecodeArrow(num_values_, null_count_, valid_bits_, 0, &builder);
+  CheckDecodeDense(actual_num_values, builder);
+}
+
+TEST_P(FromDictEncoding, DecodeNonNullDense) {
+  // Skip this test if input data contains nulls (optionals)
+  SKIP_TEST_IF(null_count_ > 0)
+  ::arrow::internal::ChunkedBinaryBuilder builder(static_cast<int>(dict_buffer_->size()),
+                                                  default_memory_pool());
+  auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder_.get());
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values = byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDense(actual_num_values, builder);
+}
+
+TEST_P(FromDictEncoding, DecodeBinaryDict) {
+  ::arrow::BinaryDictionaryBuilder builder(default_memory_pool());
+  auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder_.get());
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values =
+      byte_array_decoder->DecodeArrow(num_values_, null_count_, valid_bits_, 0, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+
+TEST_P(FromDictEncoding, DecodeStringDict) {
+  ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+  auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder_.get());
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values =
+      byte_array_decoder->DecodeArrow(num_values_, null_count_, valid_bits_, 0, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+
+TEST_P(FromDictEncoding, DecodeNonNullBinaryDict) {
+  // Skip this test if input data contains nulls (optionals)
+  SKIP_TEST_IF(null_count_ > 0)
+  ::arrow::BinaryDictionaryBuilder builder(default_memory_pool());
+  auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder_.get());
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values = byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+TEST_P(FromDictEncoding, DecodeNonNullStringDict) {
+  // Skip this test if input data contains nulls (optionals)
+  SKIP_TEST_IF(null_count_ > 0)
+  ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+  auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder_.get());
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values = byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+
+const char* json[] = {"[\"foo\", \"bar\", \"foo\", \"foo\"]",
+                      "[\"foo\", \"bar\", \"foo\", null]"};
```

Comment:
Can you generate general richer random examples? These ones aren't very interesting